R/nowcast_list_to_df.R

```diff
@@ -0,0 +1,108 @@
+#' Generate probabilistic nowcast dataframe
+#'
+#' This function ingests a list of matrices that represent expected observed
+#'    nowcasts, and generates a long tidy dataframe indexed by time and delay.
+#'
+#' @param list_of_nowcasts List of  matrices of expected
+#'    observed reporting squares.
+#' @importFrom cli cli_abort
+#' @returns `nowcast_df` Dataframe containing observations and expected
+#'    observed nowcasts indexed by reference time and delay.
+#' @export
+#'
+#' @examples
+#' point_nowcast <- matrix(
+#'   c(
+#'     80, 50, 25, 10,
+#'     100, 50, 30, 20,
+#'     90, 45, 25, 17.8,
+#'     80, 40, 23.2, 15.9,
+#'     70, 35, 20.3, 19.9
+#'   ),
+#'   nrow = 5,
+#'   byrow = TRUE
+#' )
+#'
+#' list_of_exp_obs_nowcast <- add_obs_errors_to_nowcast(
+#'   comp_rep_square = point_nowcast,
+#'   disp = c(8, 1.4, 4),
+#'   n_draws = 10
+#' )
+#'
+#' nowcast_df <- nowcast_list_to_df(
+#'   list_of_nowcasts = list_of_exp_obs_nowcast
+#' )
+nowcast_list_to_df <- function(list_of_nowcasts) {
+  # Check if all elements are matrices
+  if (!all(sapply(list_of_nowcasts, is.matrix))) {
+    cli_abort("All elements in the list must be matrices.")
+  }
+  if (length(list_of_nowcasts) < 1) {
+    cli_abort("List of nowcasts is empty")
+  }
+
+  # Convert each matrix to a data frame with an index column
+  combined_df <- do.call(
+    rbind,
+    lapply(
+      seq_along(list_of_nowcasts),
+      function(i) {
+        return(convert_reporting_square_to_df(
+          list_of_nowcasts[[i]],
+          draw = i
```

Comment:
This assumes that the intake is a probabilistic sample rather than say a list by reference date or any other grouping. In a new issue it would be nice to think about generalising this as it isn't reflected in the docs and I think is overly restrictive.

Reply:
Made a new issue for this, will flesh out further depending on response here. 

Perhaps the option `draw` argument should be just a more general way of tracking the items in the list, and then in the `nowcast_list_to_df` we convert that column to indicate that in this case it is a draw?

Reply:
>  returns a dataframe with a column for draw (so it does assume intake is list of draws)...

yes this is the problem